android/src/main/java/com/wultra/android/malwarelytics/reactnative/MalwarelyticsNativeHelper.kt

```diff
@@ -0,0 +1,210 @@
+     */
+    internal fun onDevelopmentReset() {
+        checkWorkerQueue()
+        activitiesRegisteredForTapjackingProtection.clear()
+    }
+
+    /**
+     * Execute action on the internal serial queue. If the current thread is already the thread
+     * associated to the queue, then execute action immediately.
+     *
+     * Note that this method doesn't handle exceptions.
+     */
+    internal fun executeOnQueue(action: () -> Unit) {
+        if (Thread.currentThread().id == serialExecutorThread) {
+            action()
+        } else {
+            serialExecutor.execute(action)
+        }
+    }
+
+    /**
+     * Execute action on main application's thread. If the current thread is already the main thread, then execute
+     * action immediately.
+     *
+     * Note that this method doesn't handle exceptions.
+     */
+    internal fun executeOnMainThread(action: () -> Unit) {
+        if (Looper.myLooper() == Looper.getMainLooper()) {
+            action()
+        } else {
+            mainThreadHandler.post(action)
+        }
+    }
+
+    /**
+     * Execute provided closure on the serial queue. If execution fails, then the exception is consumed and reported to the console.
+     * @param action Action to execute on the serial queue.
+     */
+    internal fun onQueue(action: (service: AppProtection) -> Unit) {
+        executeOnQueue {
+            try {
+                action(appProtectionService)
+            } catch (e: Throwable) {
+                printError("Background task failed: $e")
+            }
+        }
+    }
+
+    /**
+     * Validate whether function is called from the internal serial queue. If this is different
+     * thread then
+     */
+    internal fun checkWorkerQueue() {
+        if (Thread.currentThread().id != serialExecutorThread) {
+            throw IllegalStateException("Method called on wrong thread")
+        }
+    }
+
+    /**
+     * Print warning to the console. The function is using [RNLog] under the hood.
+     */
+    internal fun printWarning(message: String) {
+        if (BuildConfig.DEBUG) {
+            RNLog.l("WARNING: $message")
+        }
+    }
+
+    /**
+     * Print error to the console. The function is using [RNLog] under the hood.
+     */
+    internal fun printError(message: String) {
+        if (BuildConfig.DEBUG) {
+            RNLog.l("ERROR: $message")
+        }
+    }
+
+    /**
+     * Single thread executor.
+     */
+    private val serialExecutor = Executors.newSingleThreadExecutor()
+
+    /**
+     * Identifier of thread used by the serial executor.
+     */
+    private var serialExecutorThread: Long = -1
+
+    /**
+     * Handler for executing tasks on main thread.
+     */
+    private val mainThreadHandler = Handler(Looper.getMainLooper())
+
+    init {
+        serialExecutor.execute {
+            // Capture the thread identifier for optimized tasks execution.
+            serialExecutorThread = Thread.currentThread().id
```

Comment:
Tricks like these can be better handled by providing `ThreadFactory` to the executor. But in this case of a single thread, it seems ok.